Hubapi documentation (#73)
jina-ai/cloud-ops: hubapi/README.MD

> `HubAPI`

Reviewer: ```suggestion
# HubAPI

```

> `HubAPI`

Reviewer 2: Make this a top-level heading?

> The HubAPI supports operations for listing jina hub images

Reviewer 2: When we're talking about Jina Hub outside of the actual command line, call it **J**ina **H**ub

On command line or code blocks use `jina hub`

> The HubAPI supports operations for listing jina hub images

Reviewer: Why are there line feeds within the paragraph?

and pushing images to the hub. These are authorised using the

> hub authorizer. We use MongoDB for storing and retrieving

Reviewer 2: Capitalize **H**ub in all cases when used in text (not code)

information related to hub. This document elaborates on the functioning
of each of these APIs

> 1. `Hub List` API

Reviewer 2: `hub list` since its a command (all commands should be exactly as you'd write on command line)

> 1. `Hub List` API

Reviewer: ```suggestion

## Hub List API

```

> List API provides the user the ability to query jina images. Under the hood, the implementation uses a Mongodb Handler to connect to the database & insert documents in the collection.

Reviewer 2: *List API provides the user the ability to query jina images*

Simplify to:

*The list API lets users query **Jina** images*

> List API provides the user the ability to query jina images. Under the hood, the implementation uses a Mongodb Handler to connect to the database & insert documents in the collection.

Reviewer: ```suggestion
List API provides the user the ability to query `Jina` images. Under the hood, the implementation uses a `MongodbHandler` to connect to the database & insert documents in the collection.
```

Make sure wrap any code-related words in ``

> Documents are generally manifest files that have metadata information of hub images. Currently, the List API is optimised for listing the images with the latest jina version.

Reviewer 2: jina -> Jina

Reviewer 2: Are they *generally* manifest files, or *always*? (I don't actually know. If always, can remove `generally` here)

> * API gateway endpoint: `/GET` `images`        

Reviewer: Shouldn't it be `GET /images`?

* Supports cursor based pagination and alphabetical sorting
* Returns only latest images for all executors (only one result per executor)
* Returns docker pull command in addition to all info in manifest file.
* Semver sort implementation with different fields for `major`, `minor` & `patch`
* Optimized filter for searching with params:
`kind` (indexer, encoder, segmenter etc),
`type` (pod, app etc),
`jina version`,
`executor (manifest) version`,
`keywords` (array search),

> `executor name` (simple regex search).

Reviewer 2: Maybe better as a table

  
The `Aggregation` pipeline has the following stages:

> 

Reviewer 2: For this section, can we break stages down into their own sub-headings?

`Stage 1:` -> `### Stage 1`

>     Stage 1:

Reviewer: This is rendered as code. Unindent it

    - Form `$match` from different query strings
    - This will be executed only if proper query strings are passed
    - Accepted query strings -
      - `name` -> simple regex search
      - `jina_version` -> exact match
      - `version` -> exact match
      - `kind` -> exact match
      - `type` -> exact match
      - `keywords` -> array search

    Stage 2:

>     - Sort by `name`, `jina_version` & `version`

Reviewer 2: `&` -> `and` (more formal, better for docs)

>     - `name`: 1 - this sorts alphabetically (set to -1 for reverse order)

Reviewer 2: Wrap things like `1` and `-1` in backticks since they're variables

    - `jina_version`: -1 - Sorted according to semver. This should always be latest first.
    - `version`: -1 - Sorted according to semver. This should also be latest first.

    Stage 3:
    - Groupby `name` of the executor
    - Add `jina_version`, `version` & `manifest_info` to `images` list (this will be ordered)

    Stage 4:
    - Since groupby removes the sort on `name`, sort it again
    
    sort_by_name_again_stage = {
        '$sort': {
            '_id.name': 1
        }

>     } 

Reviewer 2: Should this be in a code block?

    Stage 5:
    - Pagination: Fetch images only after a certain image name

    Stage 6:

>     - Fetch n results at a time

Reviewer 2: *n* - if it's just a number placeholder
`n` - if we're referring to an actual variable

    Stage 7:

>      - 1st element of the `images` list would be the latest executor

Reviewer 2: 1st -> First

    Stage 8:

>     - Finalize the set of arguments to be set

Reviewer 2: Finalize the arguments to be set

> 2. `Hub Push` API

Reviewer: ```suggestion

## Hub Push API

```
